src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +432,640 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+    · exact hrd
+    · exact hd
+  · apply BitVec.eq_of_toNat_eq
+    rw [toNat_umod]
+    replace hdqnr : (d.toNat * q.toNat + r.toNat) % d.toNat = n.toNat % d.toNat := by
+      simp [hdqnr]
+    rw [Nat.add_mod, Nat.mul_mod_right] at hdqnr
+    simp at hdqnr
+    replace hrd : r.toNat < d.toNat := by
+      rw [BitVec.lt_def] at hrd
+      exact hrd -- TODO: golf
+    rw [Nat.mod_eq_of_lt hrd] at hdqnr
+    simp [hdqnr]
+
+theorem div_characterized_of_mul_add_of_lt {d n q r : BitVec w} (hd : 0 < d)
+    (hrd : r < d)
+    (hdqnr : d * q + r = n)
+    (hlt : d.toNat * q.toNat + r.toNat < 2^w) :
+    (n.udiv d = q ∧ n.umod d = r) := by
+  apply div_characterized_of_mul_add_toNat <;> try assumption
+  apply Eq.symm
+  have hlt' : d.toNat * q.toNat < 2^w := by omega
+  calc
+    n.toNat = (d * q + r).toNat := by rw [← hdqnr]
+    _ = ((d * q).toNat + r.toNat) % 2^w := by simp [BitVec.toNat_add]
+    _ = ((d.toNat * q.toNat) % 2^w + r.toNat) % 2^w := by simp [BitVec.toNat_mul]
+    _ = ((d.toNat * q.toNat) + r.toNat) % 2^w := by simp [Nat.mod_eq_of_lt hlt']
+    _ = ((d.toNat * q.toNat) + r.toNat)  := by simp [Nat.mod_eq_of_lt hlt]
+
+theorem div_characterized_toNat_of_eq_udiv_of_eq_umod {d n q r : BitVec w} (hd : 0 < d)
+    (hq : n.udiv d = q) (hr : n.umod d = r) :
+    (d.toNat * q.toNat + r.toNat = n.toNat) := by
+  have hdiv : n.toNat / d.toNat = q.toNat := by
+    rw [← toNat_udiv] -- TODO: squeeze
+    rw [(toNat_eq _ _).mp hq]
+  have hmod : n.toNat % d.toNat = r.toNat := by
+    rw [← toNat_umod] -- TODO: squeeze
+    rw [(toNat_eq _ _).mp hr]
+  rw [← hdiv, ← hmod] -- TODO: flip
+  rw [div_add_mod]
+
+theorem div_characterized_toNat_of_eq_udiv_of_eq_umod_of_lt {d n q r : BitVec w} (hd : 0 < d)
+    (hq : n.udiv d = q) (hr : n.umod d = r)
+    (hlt : d.toNat * q.toNat + r.toNat < 2^w) :
+    d * q + r = n := by
+  apply eq_of_toNat_eq
+  simp [toNat_add, toNat_mul]
+  rw [Nat.mod_eq_of_lt hlt]
+  apply div_characterized_toNat_of_eq_udiv_of_eq_umod hd hq hr
+
+theorem div_iff_add_mod_of_lt {d n q r : BitVec w} (hd : 0 < d)
+    (hrd : r < d)
+    (hlt : d.toNat * q.toNat + r.toNat < 2^w) :
+    (n.udiv d = q ∧ n.umod d = r) ↔ (d * q + r = n) := by
+  constructor
+  · intros h; obtain ⟨h₁, h₂⟩ := h
+    apply div_characterized_toNat_of_eq_udiv_of_eq_umod_of_lt <;> assumption
+  · intros h
+    apply div_characterized_of_mul_add_of_lt <;> assumption
+
+/-# Tons of Lemmas for Proving Bitblasting Correct -/
+
+theorem shiftLeft_eq_mul_twoPow (x : BitVec w) (n : Nat) :
+    x <<< n = x * (BitVec.twoPow w n) := by
+  ext i
+  simp
+
+/-- If `y ≤ x`, then `x - y` does not overflow, and `(x - y).toNat = x.toNat - y.toNat`. -/
+theorem toNat_sub_eq_toNat_sub_toNat_of_le {x y : BitVec w} (h : y ≤ x) :
+    (x - y).toNat = x.toNat - y.toNat := by
+  rw [BitVec.le_def] at h
+  simp only [toNat_sub]
+  simp only [show 2 ^ w - y.toNat + x.toNat = 2 ^ w + (x.toNat - y.toNat) by omega, add_mod_left]
+  rw [Nat.mod_eq_of_lt (by omega)]
```

Comment:
These belong in the general Lemmas file

Reply:
The next few lemmas as well, please check all of them

Reply:
Moved.